drv/lpc55-spi-server/src/main.rs

```diff
@@ -63,74 +369,339 @@ fn main() -> ! {
+
+            // Messages from the RoT to SP do not begin writing to FIFOWR
+            // until the SPI bus is idle (CSn not asserted).
+            // next CSn is asserted.
+            if !inframe && tctx.state == TxState::Queued {
+                tctx.state = TxState::Writing;
+                spi.txerr_clear();
+                spi.enable_tx();
+                ringbuf_entry!(Trace::StartTx(ssa, ssd));
+                gpio.set_val(rot_irq, Value::Zero).unwrap_lite();
+                tctx.rot_irq = true;
+                ringbuf_entry!(Trace::AssertRotIrq);
+            }
+
+            if tctx.state == TxState::Writing {
+                loop {
+                    if !spi.can_tx() {
+                        break;
+                    }
+                    let b = tctx.tx[tctx.count];
+                    ringbuf_entry!(Trace::Tx(tctx.state, b, tctx.count));
+                    spi.send_u8(b);
+                    tctx.count += 1;
+                    if tctx.count == tctx.end {
+                        tctx.state = TxState::Finish;
+                        break;
+                    }
+                }
+                if tctx.state == TxState::Writing {
+                    // There are more bytes to transmit
+                    again = true; // XXX needed?
                 }
-                again = true;
             }
 
-            if spi.has_byte() && !rx_done {
-                let b = spi.read_u8();
-                ringbuf_entry!(Trace::Rx(b));
-                rx[rx_cnt] = b;
-                rx_cnt += 1;
-                if rx_cnt == rx.len() {
-                    rx_done = true;
+            // TODO: check for tx underflow.
+            // TODO: check for rx overflow.
+            // TODO: Queue a failure message on those conditions.
+            // TODO: Clear the error condition.
+
+            // If we are asserting ROT_IRQ, check for conditions to de-assert.
+            if tctx.rot_irq && tctx.state == TxState::Finish {
+                let (_, _, _, txempty, _, _, _, _, _) = spi.stat();
+
+                if txempty {
+                    ringbuf_entry!(Trace::Line);
+                    // Note: SP is going to be edge-triggered, so this signal
+                    // does not need to be asserted for the entire transaction.
+                    // However, it is nice to see it on the logic analyzer for
+                    // the whole frame.
+                    gpio.set_val(rot_irq, Value::One).unwrap_lite();
+                    ringbuf_entry!(Trace::DeassertRotIrq);
+                    tctx.rot_irq = false;
+                    tctx.state = TxState::Idle;
+                    ringbuf_entry!(Trace::DeassertRotIrq);
+                    // If we have nothing to send in the next frame, then
+                    // the LPC55 will send a byte of unknown origin.
+                    spi.send_u8(0); // write our no-op byte.
+                    spi.drain_tx(); // Throw away byte just written.
```

Comment:
This seems to be a quirk of the LPC55 SPI hardware.
If the FIFO does not have some last explicit contents, it will clock out what seems to be garbage.
This is visible on a logic analyzer, but we may also find some corner case where the SP is clocking out data
and LPC55 is out of sync. Rather than waste time, (like I did when discovering this quirk), default to zeros.